File: src/main/java/org/cascadebot/cascadebot/commands/core/StatsCommand.java
package org.cascadebot.cascadebot.commands.core;

import com.sun.management.OperatingSystemMXBean;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Member;
import org.apache.commons.io.FileUtils;
import org.cascadebot.cascadebot.CascadeBot;
import org.cascadebot.cascadebot.commandmeta.CommandContext;
import org.cascadebot.cascadebot.commandmeta.ICommandCore;
import org.cascadebot.cascadebot.messaging.MessagingObjects;
import org.cascadebot.cascadebot.music.MusicHandler;

import java.lang.management.ManagementFactory;

public class StatsCommand implements ICommandCore {

    @Override
    public void onCommand(Member sender, CommandContext context) {
        EmbedBuilder builder = MessagingObjects.getClearThreadLocalEmbedBuilder();
        OperatingSystemMXBean osBean = ManagementFactory.getPlatformMXBean(OperatingSystemMXBean.class);

        builder.setTitle(context.getSelfUser().getName());
        builder.setThumbnail(context.getSelfUser().getAvatarUrl());

        builder.addField("Total Guilds", String.valueOf(CascadeBot.INS.getShardManager().getGuilds().size()), true);
        builder.addField("Active Guilds", String.valueOf(CascadeBot.INS.getShardManager().getGuildCache().size()), true);
        builder.addField("Active Voice Channels", String.valueOf(MusicHandler.getPlayers().entrySet().stream().filter(entry -> entry.getValue().getConnectedChannel() != null).count()), true);
        builder.addField("RAM Usage", FileUtils.byteCountToDisplaySize(Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()), true);
        builder.addField("CPU Load", Math.round(osBean.getProcessCpuLoad() * 100) + "%", true);
        builder.addField("Total Shards", String.valueOf(CascadeBot.INS.getShardManager().getShardsTotal()), true);
        builder.addField("Online Shards", String.valueOf(CascadeBot.INS.getShardManager().getShardsRunning()), true);
        builder.addField("Gateway Ping", String.valueOf(context.getChannel().getJDA().getGatewayPing()), true);
        builder.addField("REST Ping", String.valueOf(context.getChannel().getJDA().getRestPing().complete()), true);
        builder.addField("Shard Status", CascadeBot.INS.getShardManager().getStatus(context.getChannel().getJDA().getShardInfo().getShardId()).toString(), true);
        builder.addField("Shard ID", String.valueOf(context.getChannel().getJDA().getShardInfo().getShardId() + 1), true);

        context.getTypedMessaging().replyInfo(builder);
    }

    @Override
    public String command() {
        return "stats";
    }

    @Override
    public String description() {
        return "Returns information regarding the Bot.";
    }

}

File: src/main/kotlin/org/cascadebot/cascadebot/commands/core/StatsCommand.kt
/*
 * Copyright (c) 2020 CascadeBot. All rights reserved.
 * Licensed under the MIT license.
 */

package org.cascadebot.cascadebot.commands.core

import com.sun.management.OperatingSystemMXBean
import net.dv8tion.jda.api.entities.Member
import org.apache.commons.io.FileUtils
import org.cascadebot.cascadebot.CascadeBot
import org.cascadebot.cascadebot.commandmeta.CommandContext
import org.cascadebot.cascadebot.commandmeta.CoreCommand
import org.cascadebot.cascadebot.messaging.MessagingObjects
import org.cascadebot.cascadebot.utils.FormatUtils
import java.lang.management.ManagementFactory
import kotlin.math.roundToInt

class StatsCommand : CoreCommand() {

    override fun onCommand(sender: Member, context: CommandContext) {

        val builder = MessagingObjects.getClearThreadLocalEmbedBuilder(context.user, context.locale)
        builder.setTitle(context.selfUser.name)
        builder.setThumbnail(context.selfUser.effectiveAvatarUrl)

        val osBean = ManagementFactory.getPlatformMXBean(OperatingSystemMXBean::class.java)

        val shardManager = CascadeBot.INS.shardManager
        val fields = mapOf(
                "total_guilds"          to shardManager.guilds.size.toString(),
                "active_guilds"         to shardManager.guildCache.size().toString(),
                "ram_usage"             to FileUtils.byteCountToDisplaySize(Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()),
                "cpu_load"              to "${(osBean.processCpuLoad * 100).roundToInt()}%",
                "total_shards"          to shardManager.shardsTotal.toString(),
                "online_shards"         to shardManager.shardsRunning.toString(),
                "gateway_ping"          to "${context.channel.jda.gatewayPing}ms",
                "rest_ping"             to "${context.channel.jda.restPing.complete()}ms",
                "shard_status"          to FormatUtils.formatEnum(shardManager.getStatus(context.channel.jda.shardInfo.shardId), context.locale),
                "shard_id"              to (context.channel.jda.shardInfo.shardId + 1).toString()
        )

        fields.forEach {
            builder.addField(context.i18n("commands.stats.embed.${it.key}"), it.value, true)
        }

        context.typedMessaging.replyInfo(builder)

    }

    override fun command(): String = "stats"
}